pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -198,6 +204,44 @@ public class Utils {
         r.addAction(new ExecutionModelAction(stages))
     }
 
+    static boolean withinAStage() {
+        CpsThread thread = CpsThread.current()
+        CpsFlowExecution execution = thread.execution
+
+        DepthFirstScanner scanner = new DepthFirstScanner();
+
+        FlowNode stageNode = scanner.findFirstMatch(execution, new Predicate<FlowNode>() {
+            @Override
+            public boolean apply(FlowNode input) {
+                return input != null &&
+                    input instanceof StepStartNode &&
+                    ((StepStartNode)input).descriptor instanceof StageStep.DescriptorImpl
+            }
+        })
+
+        return stageNode != null
+    }
+
+    /**
+     * Marks the containing stage with this name as a synthetic stage, with the appropriate context.
+     *
+     * @param stageName
+     * @param context
+     */
+    static void markSyntheticStage(String stageName, String context) {
+        CpsThread thread = CpsThread.current()
+        CpsFlowExecution execution = thread.execution
+
+        FlowNode currentNode = execution.currentHeads.find { n ->
+            n?.displayName?.equals(stageName)
+        }
+
+        if (currentNode.getAction(TagsAction.class) == null) {
```

Comment:
also note: it needs to be exactly this way.  If you add the empty action first, you're persisting to disk 2x -- that's expensive, don't do it.